src/server.c

```diff
@@ -641,20 +641,19 @@ int htNeedsResize(dict *dict) {
  * In non cluster-enabled setup, it resize main/expires dictionary based on the same condition described above. */
 void tryResizeHashTables(int dbid) {
     redisDb *db = &server.db[dbid];
-    int slot = 0;
     for (dbKeyType subdict = DB_MAIN; subdict <= DB_EXPIRES; subdict++) {
-        dbIterator *dbit = dbIteratorInitFromSlot(db, subdict, db->sub_dict[subdict].resize_cursor);
-        for (int i = 0; i < CRON_DBS_PER_CALL; i++) {
-            dict *d = dbGetDictFromIterator(dbit);
-            slot = dbIteratorGetCurrentSlot(dbit);
-            dbIteratorNextDict(dbit);
-            if (!d) break;
+        if (dbSize(db, subdict) == 0) continue;
+
+        if (db->sub_dict[subdict].resize_cursor == -1)
+            db->sub_dict[subdict].resize_cursor = findSlotByKeyIndex(db, 1, subdict);
+
+        for (int i = 0; i < CRON_DBS_PER_CALL && db->sub_dict[subdict].resize_cursor != -1; i++) {
+            int slot = db->sub_dict[subdict].resize_cursor;
+            dict *d = subdict == DB_MAIN ? db->dict[slot] : db->expires[slot];
```

Comment:
Shall we move it to the outer loop?

Reply:
As the "slot" is a variable within the loop, it cannot be moved outside.

Maybe you mean something like this?
```c
dict *dbDict(redisDB *db, dbKeyType keyType, int slot) {
    return keyType == DB_MAIN ? db->dict[slot] : db->expires[slot];
}

dict *d = dbDict(db, subdict, slot);
```

Reply:
Right, I'm blind 🙈

Reply:
I think we can avoid the refactoring it's pretty simple.